### PR TITLE
refactor: make code for decoding more readable

### PR DESCRIPTION
**Pull request by erikgb**
This PR prepares for splitting the `zapHook` func in a follow-up PR.

Notable changes: Removing a custom interface, and replacing it with the corresponding interface from Golang core - also checking that it can be type-casted.

### CODE REVIEW

File: pkg/zap/zap.go [hunk 1]
```diff
@@ -133,60 +134,61 @@ var levelStrings = map[string]zapcore.Level{
...
-				return encoder, nil
-			}
+			// TODO: implement encoding.TextUnmarshaler interface for type NewEncoderFunc upstream
+			var encoder crzap.NewEncoderFunc
 
-			switch encoderIn {
+			switch val.(string) {
+			case "": // Encoder not configured; use default encoder
+				// TODO: Isn't the default encoder dependant on the development/production setting?
+				encoder = newJSONEncoder
```
Comment (padlar) [inline pkg/zap/zap.go:167]: good catch! https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/log/zap/zap.go#L185 looking at the addDefaults logic, NewEncoder is set correctly based on ZAP_DEVELOPMENT flag, so I think we should set encoder to nil here

Comment (erikgb) [reply pkg/zap/zap.go:167]: I think we should create an issue to follow up on this. If setting it to nil, the test fails:

````
Summarizing 1 Failure:
  [FAIL] Zap log level flag options setup with zap-encoding flag provided [It] Should default to json encoder when not set [encoder]
  /home/erikbo/projects/github/controller-runtime-viper/pkg/zap/zap_test.go:241
````

Comment (padlar) [reply pkg/zap/zap.go:167]: Opened issue #51

**Review by padlar — APPROVED** [review]
lgtm

**Pull request merged**